From 077a12c0c2187bcb24cec6bcd1e9918a6b97c124 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Wed, 17 Jan 2024 18:34:34 +0100
Subject: [PATCH 481/954] refactor(stm32mp2): revisit
 stm32mp2_syscfg_get_mm_size

This change saves a few bytes of code.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 8af4d86e477d43facb668951801abb282cb8c042 refactor(stm32mp2): revisit stm32mp2_syscfg_get_mm_size]
Change-Id: I14c4c584a1c59a2f61bfd2188c01fcef75546f54
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/352187
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_syscfg.c | 32 ++++++++++--------------------
 1 file changed, 10 insertions(+), 22 deletions(-)

diff --git a/plat/st/stm32mp2/stm32mp2_syscfg.c b/plat/st/stm32mp2/stm32mp2_syscfg.c
index 9c4c63c9eb..aca04ffcbc 100644
--- a/plat/st/stm32mp2/stm32mp2_syscfg.c
+++ b/plat/st/stm32mp2/stm32mp2_syscfg.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2023-2024, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -136,30 +136,18 @@ uint32_t stm32mp_syscfg_get_chip_dev_id(void)
 size_t stm32mp_syscfg_get_mm_size(uint8_t bank)
 {
 	uint32_t amcr = mmio_read_32(SYSCFG_BASE + SYSCFG_OCTOSPIAMCR);
-	size_t addr_mapping1 = 0U;
-	size_t addr_mapping2 = 0U;
-
-	switch (amcr & SYSCFG_OCTOSPIAMCR_OAM_MASK) {
-	case 0:
-		addr_mapping1 = SZ_256M;
-		break;
-	case 1:
-		addr_mapping1 = SZ_128M + SZ_64M;
-		addr_mapping2 = SZ_64M;
-		break;
-	case 2:
-		addr_mapping1 = SZ_128M;
-		addr_mapping2 = SZ_128M;
-		break;
-	case 3:
-		addr_mapping1 = SZ_64M;
-		addr_mapping2 = SZ_128M + SZ_64M;
-		break;
-	default:
+	uint32_t oam = amcr & SYSCFG_OCTOSPIAMCR_OAM_MASK;
+	size_t addr_mapping1;
+	size_t addr_mapping2;
+
+	if (oam > 3U) {
 		addr_mapping2 = SZ_256M;
-		break;
+	} else {
+		addr_mapping2 = oam * SZ_64M;
 	}
 
+	addr_mapping1 = SZ_256M - addr_mapping2;
+
 	return (bank == 0U) ? addr_mapping1 : addr_mapping2;
 }
 
-- 
2.53.0

